matrix/static_matrix.h

```diff
@@ -0,0 +1,888 @@
+    static_matrix<T, Rows, Cols> &operator*=(static_matrix<T, Rows, Cols> &lhs, const U &value) {
+        static_assert(std::is_convertible<U, T>::value, "U must be convertible to T");
+#endif // C++ <= 201703L
+        lhs.mul(value);
+        return lhs;
+    }
+
+#if __cplusplus > 201703L
+    template<fundamental T, fundamental U, std::size_t Rows, std::size_t Cols>
+    requires (std::convertible_to<U, T>)
+    MATRIX_CXX17_NODISCARD MATRIX_CXX17_CONSTEXPR static_matrix<T, Rows, Cols> operator/(const static_matrix<T, Rows, Cols> &lhs, const U &value) {
+#else
+    template<typename T, typename U, std::size_t Rows, std::size_t Cols>
+    MATRIX_CXX17_NODISCARD MATRIX_CXX17_CONSTEXPR static_matrix<T, Rows, Cols> operator/(const static_matrix<T, Rows, Cols> &lhs, const U &value) {
+        static_assert(std::is_convertible<U, T>::value, "U must be convertible to T");
+#endif // C++ <= 201703L
+        static_matrix<T, Rows, Cols> division = lhs.div_by_element(static_matrix<T, Rows, Cols>(value));
+        return division;
+    }
+
+#if __cplusplus > 201703L
+    template <fundamental T, fundamental U, std::size_t Rows, std::size_t Cols>
+    requires (std::convertible_to<U, T>)
+    static_matrix<T, Rows, Cols> &operator/=(static_matrix<T, Rows, Cols>&lhs, const U &value) {
+#else
+    template <typename T, typename U, std::size_t Rows, std::size_t Cols>
+    static_matrix<T, Rows, Cols> &operator/=(static_matrix<T, Rows, Cols>&lhs, const U &value) {
+        static_assert(std::is_convertible<U, T>::value, "U must be convertible to T");
+#endif // C++ <= 201703L
+        lhs.div(value);
+        return lhs;
+    }
+
+#if __cplusplus > 201703L
+    template <fundamental T, std::size_t Rows, std::size_t Cols>
+    MATRIX_CXX17_NODISCARD MATRIX_CXX17_CONSTEXPR bool operator==(const static_matrix<T, Rows, Cols> &lhs, const static_matrix<T, Rows, Cols> &rhs) {
+#else
+    template <typename T, std::size_t Rows, std::size_t Cols>
+    MATRIX_CXX17_NODISCARD MATRIX_CXX17_CONSTEXPR bool operator==(const static_matrix<T, Rows, Cols> &lhs, const static_matrix<T, Rows, Cols> &rhs) {
+#endif // C++ <= 201703L
+        return lhs.equal_to(rhs);
+    }
+
+#if __cplusplus > 201703L
+    template <fundamental T, std::size_t Rows, std::size_t Cols>
+    MATRIX_CXX17_NODISCARD MATRIX_CXX17_CONSTEXPR bool operator!=(const static_matrix<T, Rows, Cols> &lhs, const static_matrix<T, Rows, Cols> &rhs) {
+#else
+    template <typename T, std::size_t Rows, std::size_t Cols>
+    MATRIX_CXX17_NODISCARD MATRIX_CXX17_CONSTEXPR bool operator!=(const static_matrix<T, Rows, Cols> &lhs, const static_matrix<T, Rows, Cols> &rhs) {
+#endif // C++ <= 201703L
+        return !(lhs == rhs);
+    }
+
+#if __cplusplus <= 201703L
+    template <typename T>
+    class static_matrix<T, 0, 0> {
+        static_assert(std::is_fundamental<T>::value, "Template parameter T must be fundamental");
+    public:
+        T determinant_laplacian() { return T{}; }
```

Comment:
https://math.stackexchange.com/questions/1913141/is-the-0-times-0-matrix-zero-times-zero-matrix-a-well-defined-concept